Improve paper key phrase checks

patrickxb opened the pull request:
I believe the problem in https://keybase.atlassian.net/browse/CORE-2940 was a mistyped paper key phrase, despite the version check passing.  I have not heard back from the user yet, so will wait to close the issue until then.

In the meantime, this PR adds better paper key parsing, as suggested by @oconnor663 in https://keybase.atlassian.net/browse/CORE-2647 to get around common typos like extra spaces.  It also checks to see if each word is in the secwords list.  And there is now a libkb.Checker for paper keys.

It also adds some debugging statements and some cleanup of global G usage that I found while digging into this issue with a test program.

r? @maxtaco @oconnor663

maxtaco commented on go/engine/common.go:
```diff
@@ -61,9 +64,15 @@ func findPaperKeys(ctx *Context, g *libkb.GlobalContext, me *libkb.User) (*keypa
 	}
+	if !paperPhrase.ValidWords() {
+		g.Log.Debug("paper phrase has invalid word(s) in it")
+		return nil, libkb.PassphraseError{Msg: "invalid word(s) in paper key phrase"}
```
Maybe a different type of error message so the front end can switch on it?

patrickxb replied:
It's nearly impossible to get this error because the checker on paper key phrase already checks this.  I just left this in for safety.

maxtaco commented on go/libkb/paperkey_phrase.go:
```diff
@@ -34,7 +34,8 @@ func MakePaperKeyPhrase(version uint8) (PaperKeyPhrase, error) {
 // NewPaperKeyPhrase converts a string into a PaperKeyPhrase.
 func NewPaperKeyPhrase(phrase string) PaperKeyPhrase {
-	return PaperKeyPhrase(strings.ToLower(phrase))
+	s := strings.Join(strings.Fields(phrase), " ")
```
what will happen if there are two spaces between two words?

maxtaco replied:
seems slightly cleaner to Trip and Tolower before going to fields.  and maybe store the `[]string{}` components in the `PaperKeyPassphrase` object, and just `Join` on `String()`?

patrickxb replied:
Fields handles multiple spaces.  I added a test for that case.

maxtaco replied:
@patrickxb did you see my second comment above?  (You replied to only the first)

patrickxb replied:
Sorry I forgot to address it earlier.  I changed the order, but for now I left `type PaperKeyPhrase string` unchanged.  

Besides being a bigger change to move to []string components, the full string with the spaces is the phrase, so seems to me to be the right thing to store in memory.  But if you feel strongly, let me know.

maxtaco commented:
some small comments above

oconnor663 commented on go/libkb/checkers.go:
```diff
@@ -79,6 +79,27 @@ var CheckKex2SecretPhrase = Checker{
+		if !p.ValidWords() {
+			G.Log.Debug("paper phrase has invalid word(s) in it")
+			return false
+		}
```
I think the valid words check should come first. The way Version() is implemented, it won't notice if the last word is garbage; it'll just hash it up and (probably) report that the version is wrong.

oconnor663 commented on go/libkb/checkers.go:
```diff
@@ -79,6 +79,27 @@ var CheckKex2SecretPhrase = Checker{
+		}
+		return true
+	},
+	Hint: "It looks like there was a typo in the paper key.  Please try again.",
```
Do we show this hint even for a real invalid version error? Is there a way to be more specific?

Also, if there was a way to report *which* word was misspelled, that would be super helpful, especially if the user actually wrote the word down wrong and doesn't know the correct spelling.

oconnor663 commented on go/libkb/checkers.go:
```diff
@@ -79,6 +79,27 @@ var CheckKex2SecretPhrase = Checker{
+		if err != nil {
+			G.Log.Debug("error getting paper key version: %s", err)
+			return false
+		}
```
Right now the only possible error here is that the key is empty string, I think. Maybe just check for that explicitly? (In addition to moving up the valid words check?)

patrickxb commented:
@oconnor663 I changed the checker significantly in order to provide custom hints.  So now it will tell you in the pinentry UI which words weren't in the list and if there was a version mismatch.  Also checks valid words first, as you suggested.

Also replaced the use of global G with a contextified version, which resulted in several file changes up the call path.

oconnor663 commented on go/libkb/passphrase_helper.go:
```diff
@@ -134,3 +143,58 @@ func DefaultPassphraseArg(allowSecretStore bool) keybase1.GUIEntryArg {
+			return fmt.Errorf("Please try again. These words are invalid: %s", strings.Join(invalids, ", "))
+		}
+		return fmt.Errorf("Please try again. This word is invalid: %s", invalids[0])
+	}
```
Thanks!

maxtaco commented:
OK I withdraw my earlier comment about the representation inside of `PaperKey`

oconnor663 commented on go/libkb/passphrase_helper.go:
```diff
@@ -134,3 +143,58 @@ func DefaultPassphraseArg(allowSecretStore bool) keybase1.GUIEntryArg {
+	if err != nil {
+		g.Log.Debug("error getting paper key version: %s", err)
+		// despite the error, just tell the user there was a typo:
+		return errors.New("It looks like there was a typo in the paper key. Please try again.")
```
I think the only way this can happen right now is that the phrase is just empty string? Might make sense to have an explicit check for that above instead of calling it a typo? But we could keep the check here too, in case new errors cases show up in the future.

patrickxb replied:
They could still enter valid words and have the version be incorrect.

oconnor663 replied:
I think that's the `version != PaperKeyVersion` check below, as opposed to this `err != nil` check?

patrickxb replied:
Oh sorry, github diff playing tricks on me.  Yes, you are correct.

oconnor663 commented:
LGTM. For what it's worth I'm totally neutral about how paperkey stores things internally, since it's easy to change later.
